packages/server-wallet/src/protocols/channel-closer.ts

```diff
@@ -39,142 +38,137 @@ export class ChannelCloser {
   public async crank(objective: DBCloseChannelObjective, response: WalletResponse): Promise<void> {
     const channelToLock = objective.data.targetChannelId;
 
-    let attemptAnotherProtocolStep = true;
+    await this.store.lockApp(channelToLock, async tx => {
+      const channel = await this.store.getChannel(channelToLock, tx);
+      if (!channel) {
+        throw new Error('Channel must exist');
+      }
+      try {
+        if (!ensureAllAllocationItemsAreExternalDestinations(channel)) {
+          return;
+        }
 
-    while (attemptAnotherProtocolStep) {
-      await this.store.lockApp(channelToLock, async tx => {
-        const protocolState = await this.getCloseChannelProtocolState(
-          this.store,
-          objective.data.targetChannelId,
-          tx
-        );
-        try {
-          let turnNum;
-          if (!ensureAllAllocationItemsAreExternalDestinations(protocolState)) {
-            attemptAnotherProtocolStep = false;
-          } else if ((turnNum = turnNumberToSign(protocolState))) {
-            await this.signState(protocolState, turnNum, tx, response);
-          } else if (shouldDefund(protocolState)) {
-            await this.defund(protocolState, tx);
-          } else if (shouldCompleteObjective(protocolState)) {
-            attemptAnotherProtocolStep = false;
-            await this.completeObjective(objective, protocolState, tx, response);
-          } else {
-            response.queueChannelState(protocolState.app);
-            attemptAnotherProtocolStep = false;
-          }
-        } catch (error) {
-          this.logger.error({error}, 'Error taking a protocol step');
-          await tx.rollback(error);
-          attemptAnotherProtocolStep = false;
+        const turnNum = turnNumberToSign(channel);
+        // Explicitely check undefined just in case we want to sign turnNum 0
+        if (turnNum !== undefined) {
+          await this.signState(channel, turnNum, tx, response);
+        }
+
+        if (await this.shouldDefund(channel, tx)) {
+          await this.defund(channel, tx);
+        }
+
+        if (await this.shouldCompleteObjective(channel, tx)) {
+          await this.completeObjective(objective, channel, tx, response);
+          return;
         }
-      });
+
+        response.queueChannel(channel);
+      } catch (error) {
+        this.logger.error({error}, 'Error taking a protocol step');
+        await tx.rollback(error);
```

Comment:
I'm not sure if we need this anymore. If the something throws an errors, the promise will reject and the transaction automatically rolls back, and the error is logged. It might(?) be better to allow `crank` to error and catch it elsewhere instead of swallowing here. It seems like any error thrown here will be developer errors, rather than data errors.

Reply:
We recently made a similar change for ledger manager cranking, moving the equivalent try/catch up into the wallet's `processLedgerQueue` function https://github.com/statechannels/statechannels/pull/3185. We could do the same thing for objectives, and put the try/catch in the `crankUntilIdle` function.

Reply:
That's a good point. Filed https://github.com/statechannels/statechannels/issues/3252.